In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [2]:
data = pd.read_csv("student_class1.csv")

In [3]:
data.head()

Roll No  S1  S2  S3  S4  S5  S6  Teachers
0        1  95  77  77  73  43  93        10
1        2  89  96  65  78  99  74        10
2        3  65  62  88  91  45  75        10
3        4  78  42  78  64  69  54        10
4        5  81  68  93  65  84  54        10

In [15]:
X = data.loc[:,["S1","S2","S3","S4","S5","Teachers"]]
y = data.loc[:,"S6"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)



### Training Our model...

In [16]:
# Linear Regresssion...
from sklearn.linear_model import LinearRegression
model1 = LinearRegression()
model1.fit(X_train,y_train)
pred1 = model1.predict(X_test)
print(pred1[0])


68.47737918811019


In [17]:
# Gradient Boosting Regressor
from sklearn.ensemble import GradientBoostingRegressor
model2 = GradientBoostingRegressor()
model2.fit(X_train,y_train)
pred2 = model2.predict(X_test)
print(pred2[0:5])


[75.01273343 80.36807484 77.81186039 73.018835   84.66763387]


In [18]:
# Bayesian Ridge
from sklearn.linear_model import BayesianRidge
model3 = BayesianRidge()
model3.fit(X_train,y_train)
pred3 = model3.predict(X_test)
print(pred3[0:5])


[69.62552865 71.57912053 70.50579972 71.57225446 71.05411544]


In [19]:
# RandomForestRegressor
from sklearn.ensemble import RandomForestRegressor
model4 = RandomForestRegressor()
model4.fit(X_train,y_train)
pred4 = model4.predict(X_test)
print(pred4[0:5])

[73.4  76.04 77.47 74.16 82.76]


In [20]:
error = mean_squared_error(pred4,y_test)
error

299.3735751515151

## Checking Accuracy of Model's.. 

In [21]:
print("Acccuracy of Linear Regression:- ",model1.score(X_test,y_test))
print("Acccuracy of Gradient Boosting Regressor:- ",model2.score(X_test,y_test))
print("Acccuracy of Bayesian Ridge:- ",model3.score(X_test,y_test))
print("Acccuracy of Random Forest Regressor:- ",model4.score(X_test,y_test))

Acccuracy of Linear Regression:-  0.03197047680259557
Acccuracy of Gradient Boosting Regressor:-  -0.058252020397741466
Acccuracy of Bayesian Ridge:-  0.014542649212469061
Acccuracy of Random Forest Regressor:-  -0.04017133729112832


# Now, Case 2:


### In this Case, we will have another class  and we  will transfer one Student from Class A to Class B, to check whether that student marks will change or not via our prediction model....

* So, First we have to update our class B
* Then we will train our model with respect to Class B
* Then, again make prediction
* Then, check for Difference in Prev value and Curr value...

In [22]:
#1. 
data2 = pd.read_csv("student_class2.csv")
df = pd.DataFrame(data)
df2 = pd.DataFrame(data2)
print("Before Transfer")
print(df2.head())


Before Transfer
   Roll No  S1  S2  S3  S4  S5  S6  Teachers
0        1  94  61  46  75  44  51        10
1        2  58  87  67  81  96  93        10
2        3  68  95  82  99  98  65        10
3        4  78  78  90  65  61  94        10
4        5  90  55  84  56  52  49        10


In [23]:
print("After Transfer")
df2.loc[0] = df.loc[0]
print(df2.head())


After Transfer
   Roll No  S1  S2  S3  S4  S5  S6  Teachers
0        1  95  77  77  73  43  93        10
1        2  58  87  67  81  96  93        10
2        3  68  95  82  99  98  65        10
3        4  78  78  90  65  61  94        10
4        5  90  55  84  56  52  49        10


In [24]:
#2.
X = data.loc[:,["S1","S2","S3","S4","S5","Teachers"]]
y = data.loc[:,"S6"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [25]:
#3. Linear Regresssion...
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train,y_train)
pred = model.predict(X_test)
print(pred[0])

68.47737918811019


### Here, we see that both previous value and current value of student 1 is :-  68.47737918811019

# Case 3:

### In this, we are going to reduce the no. of teachers to 5, then predict our values...

In [26]:
print("Before Changing")
print(df2.head())

Before Changing
   Roll No  S1  S2  S3  S4  S5  S6  Teachers
0        1  95  77  77  73  43  93        10
1        2  58  87  67  81  96  93        10
2        3  68  95  82  99  98  65        10
3        4  78  78  90  65  61  94        10
4        5  90  55  84  56  52  49        10


In [28]:
print("After Changing")
df2.loc[:,"Teachers"] = 5

print(df2.head())

After Changing
   Roll No  S1  S2  S3  S4  S5  S6  Teachers
0        1  95  77  77  73  43  93         5
1        2  58  87  67  81  96  93         5
2        3  68  95  82  99  98  65         5
3        4  78  78  90  65  61  94         5
4        5  90  55  84  56  52  49         5


In [29]:
X = data.loc[:,["S1","S2","S3","S4","S5","Teachers"]]
y = data.loc[:,"S6"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [30]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train,y_train)
pred = model.predict(X_test)
print(pred[0])

68.47737918811019


### Here, we see that still the predicted current value of student 1 is :-  68.47737918811019

                                     ----- The End -----